src/state/selectors/axisSelectors.ts

```diff
@@ -1133,3 +1143,131 @@ export const selectYAxisSize = createSelector(
     };
   },
 );
+
+export const selectDuplicateDomain = createSelector(
+  selectChartLayout,
+  selectAllAppliedValues,
+  selectAxisSettings,
+  pickAxisType,
+  (chartLayout, appliedValues, axis, axisType) => {
+    if (axis == null) {
+      return undefined;
+    }
+    const { allowDuplicatedCategory, type, dataKey } = axis;
+    const isCategorical = isCategoricalAxis(chartLayout, axisType);
+    const allData = appliedValues.map(av => av.value);
+    if (dataKey && isCategorical && type === 'category' && allowDuplicatedCategory && hasDuplicate(allData)) {
+      return allData;
+    }
+    return undefined;
+  },
+);
+
+export const selectCategoricalDomain = createSelector(
+  selectChartLayout,
+  selectAllAppliedValues,
+  selectAxisSettings,
+  pickAxisType,
+  (layout, displayedData, axis, axisType) => {
+    if (axis == null) {
+      return undefined;
+    }
+    const { type, scale } = axis;
+    const isCategorical = isCategoricalAxis(layout, axisType);
+    if (isCategorical && (type === 'number' || scale !== 'auto')) {
+      return displayedData.map(d => d.value);
+    }
+    return undefined;
+  },
+);
+
+export const selectTicksOfAxis = createSelector(
+  selectChartLayout,
+  selectAxisSettings,
+  selectAxisScale,
+  selectNiceTicks,
+  selectAxisRange,
+  selectDuplicateDomain,
+  selectCategoricalDomain,
+  pickAxisType,
+  (
+    layout,
+    axis,
+    scaleReturn,
+    niceTicks,
+    axisRange,
+    duplicateDomain,
+    categoricalDomain,
+    axisType,
+  ): ReadonlyArray<CartesianTickItem> | null => {
+    if (axis == null || scaleReturn == null) {
+      return null;
+    }
+
+    const isCategorical = isCategoricalAxis(layout, axisType);
+
+    const { type, ticks, tickCount } = axis;
+
+    const { scale, realScaleType } = scaleReturn;
+    if (scale == null) {
+      return null;
+    }
+
+    const offsetForBand = realScaleType === 'scaleBand' ? scale.bandwidth() / 2 : 2;
+    let offset = type === 'category' && scale.bandwidth ? scale.bandwidth() / offsetForBand : 0;
+
+    offset =
+      axisType === 'angleAxis' && axisRange?.length >= 2 ? mathSign(axisRange[0] - axisRange[1]) * 2 * offset : offset;
+
+    // The ticks set by user should only affect the ticks adjacent to axis line
+    if (ticks || niceTicks) {
+      const result = (ticks || niceTicks).map((entry: AxisTick): CartesianTickItem => {
+        const scaleContent = duplicateDomain ? duplicateDomain.indexOf(entry) : entry;
+
+        return {
+          // If the scaleContent is not a number, the coordinate will be NaN.
+          // That could be the case for example with a PointScale and a string as domain.
+          coordinate: scale(scaleContent) + offset,
+          value: entry,
+          // @ts-expect-error why does the offset go here? The type does not require it
+          offset,
```

Comment:
Maybe for info provided to a callback event? Idk just making something up lol

Reply:
Maybe the type needs a change, dunno. I left it like this for now.